torch/nn/parallel/distributed.py

```diff
@@ -654,6 +657,30 @@ def _register_comm_hook(self, state: object, hook: callable):
             >>>     fut.set_result(bucket.get_tensors())
             >>>     return fut
 
+        Example::
+            Below is an example of a simple allreduce hook.
+
+            >>> ddp._register_comm_hook(state = None, hook = allreduce)
+
+            >>> def allreduce(state: object, bucket: dist.GradBucket): -> torch.futures.Future
+            >>>     work = dist.allreduce(bucket.get_tensors())
+            >>>     return work.get_future()
+
+        Example::
+            Below is an example of a Parallel SGD algorithm where gradients are encoded before
+            allreduce, and then decoded after allreduce.
+
+            >>> ddp._register_comm_hook(state = None, hook = encode_and_decode)
+
+            >>> def encode_and_decode(state: object, bucket: dist.GradBucket): -> torch.futures.Future
+            >>>     encoded_tensors = encode(bucket.get_tensors()) # encode gradients
+            >>>     fut = process_group.allreduce(encoded_tensors).get_future()
```

Comment:
I guess `encoded_tensors` here is not appropriate since `allreduce` takes only one tensor? We should probably do an allreduce for each tensor separately? Same for the allreduce example hook above.

Reply:
I think `process_group.allreduce` is fine because it can get a list of Tensors. The problem is with `dist.all_reduce` which just gets Tensors and not list of Tensors. 

@mrshenli also mentioned renaming this function as `get_replicas` for better engineering. Maybe we have both, and `get_tensors` returns Tensors but raises exception if there are multiple replicas? 
My plan is to address the minor API changes (like this example and skipping division by worldsize) in a stacked diff.

Reply:
Please see the new `get_tensor` method in GradBucket. I think it makes sense:
https://github.com/pytorch/pytorch/blob/f421853b0b950c2b0bae0100558f39bc3082d8c0/torch/csrc/distributed/c10d/comm.cpp#L90-L96

Reply:
I think we should have only one. I'm fine with calling it `get_replicas`, but the docs should clearly mention that each tensor in the list refers to the replica on each device. Also, mention that we will have multiple replicas only in the case of single process multiple device mode. In the single process single device mode, this list would consist of only a single tensor.

Reply:
I'll follow the recommended change here, but I didn't get how to `allreduce` each tensor separately. How can I combine those works into one Future?

Reply:
Reverted `getTensor`